Adds a v1a1 to v1a2 comparison document (#1211)
kubernetes-sigs/cluster-api: docs/developer/v1alpha1-compared-to-v1alpha2.md

# Cluster API v1alpha1 compared to v1alpha2

## Providers

### v1alpha1

Providers in v1alpha1 are generic cloud providers including, aws, gcp, openstack, azure, vsphere and others. The provider 
vendors in Cluster API's controllers, registers their actuators with the controllers and runs a single manager 

> (with the Cluster API controllers) to complete a Cluster API management cluster.

Reviewer: I'd probably call out that there is only a singular type of provider in this section and that provider is responsible for both infrastructure management and kubernetes bootstrapping.

Reviewer: It might also be good to point out here that provider-specific configuration is embedded within the Cluster and Machine types as `Spec.ProviderSpec` and provider-specific status is embedded as `Status.ProviderStatus`

Reviewer 2: The latter suggestion to specify field types seems like too much detail. The aim for this doc should be an overview of the differences.

### v1alpha2

> v1alpha2 introduces new providers. Cluster API's controllers are no longer vendored by providers. Instead, Cluster API 

Reviewer 2: I'd probably move this section at the top:
v1alpha introduces 2 new providers (infrastructure and bootstrap provider). This means that in order to have a complete management cluster that is ready to build clusters you will need three managers: infrastructure, bootstrap, and cluster api. 

Everything else ->

Author: I organized it this way so i could, briefly, define what a provider is. That allows me to make a comparison in this section. Without the definition in v1a1 I was thinking this might be a little confusing.

What are your thoughts about why you think this section should move to the top?

Reviewer 2: *To the top of v1alpha2 section. Sorry I wasn't clear.
As for why to the top of v1alpha2, it just sets up the section's context better so that I have an overview.

Author: restructured, thanks for the feedback

# Cluster API v1alpha1 compared to v1alpha2

## Providers

### v1alpha1

Providers in v1alpha1 wrap behavior specific to an environment to create the infrastructure and bootstrap instances into

> Kubernetes nodes. Examples of environments that have integrated with Cluster API v1alpha1 include, aws, gcp, openstack,

Reviewer 3: ```suggestion
Kubernetes nodes. Examples of environments that have integrated with Cluster API v1alpha1 include AWS, GCP, OpenStack,
```

> azure, vsphere and others. The provider vendors in Cluster API's controllers, registers their actuators with the

Reviewer 3: ```suggestion
Azure, vSphere, and others. The provider vendors in Cluster API's cluster and machine controllers, registers its actuators with the
```

> controllers and runs a single manager (with the Cluster API controllers) to complete a Cluster API management cluster.

Reviewer 3: ```suggestion
controllers, and runs a single manager (with the Cluster API cluster and machine controllers) to complete a Cluster API management cluster.
```

Reviewer 3: Should we mention that there are other controllers from Cluster API that providers usually don't vendor, and that are run from the cluster-api manager pod? (MachineDeployment, MachineSet, NodeRef)

Author: I'm a little torn, I omitted that detail since I don't think it's terribly important to understand the differences since it's not really different now, but definitely interested to hear why you think we should include it

Reviewer 3: If this is about differences, I think it's fine to omit.

### v1alpha2

v1alpha2 introduces new providers. Cluster API's controllers are no longer vendored by providers. Instead, Cluster API 
offers its own independent controllers that are responsible for the core types:

* Cluster
* Machine
* MachineSet
* MachineDeployment
* [plus a few more](../../pkg/apis/cluster/v1alpha2)

To complete a management cluster in v1alpha2, the Cluster API controllers need support from two other providers. A Bootstrap
Provider and an Infrastructure provider.

The Bootstrap provider is responsible for running a controller that generates data necessary to bootstrap an instance into a 
Kubernetes node (cloud-init, bash, etc).

The Infrastructure provider is responsible for generating actual infrastructure (networking, load balancers, instances, 
etc) for a particular environment that can consume bootstrap data to turn the infrastructure into a Kubernetes cluster.

This means that in order to have a complete management cluster that is ready to build clusters you will need three managers.

> v1alpha1 "providers" have become Infrastructure providers. Cluster API has moved from being a library to vendor to its

Reviewer 3: For v1alpha1, it was half library, half its own proper controllers.

Author: i don't think it's necessary to get into the details of exactly what cluster-api was, thinking about cluster api as a library seems sufficient to me for the purpose of this document.

Do you think there's a benefit to getting into the details about how capi was half library half independent controllers? Like, does omitting that detail really harm the understanding?

Reviewer 3: I think this is fine, thanks.

own set of controllers to be used on every Cluster API management cluster. Bootstrap providers are an entirely new concept
aimed at reducing the amount of kubeadm boilerplate that every provider reimplemented in v1alpha1.

## Actuators

### v1alpha1

Actuators are interfaces that the Cluster API controller calls. A provider pulls in the generic Cluster API controller
and then registers actuators to run specific infrastructure logic (calls to the provider cloud).
 
### v1alpha2

Actuators are not used in this version. Cluster API's controllers are no longer shared across providers and therefore
they do not need to know about the actuator interface. Instead, providers communicate to each other via Cluster API's

> central objects, namely Machine and Cluster. When a user modifies a Machine with a reference each provider will notice

Reviewer 3: ```suggestion
central objects, namely Machine and Cluster. When a user modifies a Machine with a reference, each provider will notice
```

> that update and respond in some way. For example, the Bootstrap provider may attach BootstrapData to a Machine or the

Reviewer 3: A bootstrap provider should only ever modify its own custom resource (to set status.bootstrapData). The Cluster API Machine Controller will pull (copy) the value from the bootstrap custom resource to the Machine.

Author: hah. this process also helps debugging my own mental model :D
